core/src/main/java/org/jdbi/v3/BuiltInArgumentFactory.java

```diff
@@ -100,31 +105,54 @@ public static Argument build(Object arg) {
     }
 
     @Override
-    public boolean accepts(Class<?> expectedType, Object value, StatementContext ctx)
+    public boolean accepts(Type expectedType, Object value, StatementContext ctx)
     {
-        return BUILDERS.containsKey(expectedType) || value == null || value.getClass().isEnum();
+        return BUILDERS.containsKey(expectedType)
+                || value == null
+                || value.getClass().isEnum()
+                || value instanceof Optional;
     }
 
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({ "unchecked", "rawtypes" })
     @Override
-    public Argument build(Class<?> expectedType, T value, StatementContext ctx)
+    public Argument build(Type expectedType, Object value, StatementContext ctx)
     {
-        Class<?> buildType = value != null && expectedType == Object.class ? value.getClass() : expectedType;
-        ArgBuilder<T> v = (ArgBuilder<T>) BUILDERS.get(buildType);
+        Class<?> expectedClass = getErasedType(expectedType);
 
+        if (value != null && expectedClass == Object.class) {
+            expectedClass = value.getClass();
+        }
+
+        ArgBuilder v = BUILDERS.get(expectedClass);
         if (v != null) {
             return v.build(value);
         }
 
         // Enums must be bound as VARCHAR.
-        if (value != null && value.getClass().isEnum()) {
+        if (value instanceof Enum) {
             return STR_BUILDER.build((((Enum<?>)value).name()));
         }
 
+        if (value instanceof Optional) {
```

Comment:
I was going by the "built in" part of the class name. There are certain things we'd expect to come built in and I would argue that Optional is one that people would expect out of the box.